### PR TITLE
Three dimensional difference operators

### PR DESCRIPTION
**Pull request by simone-silvestri**
instead of 3D conditional derivatives, this PR introduces 3D conditional differences (and as a result derivatives)

### CODE REVIEW

**Review by glwagner — APPROVED** [review]
Makes sense

File: src/ImmersedBoundaries/conditional_derivatives.jl
Comment (navidcy) [inline src/ImmersedBoundaries/conditional_derivatives.jl:1]: rename the file to `conditional_differences.jl`?

File: src/ImmersedBoundaries/conditional_derivatives.jl [hunk 1]
```diff
@@ -1,28 +1,28 @@
...
-    ∂zᶜᶜᶠ, ∂zᶜᶠᶠ, ∂zᶠᶜᶠ, ∂zᶠᶠᶠ,
-    ∂zᶜᶜᶜ, ∂zᶜᶠᶜ, ∂zᶠᶜᶜ, ∂zᶠᶠᶜ
+    δxᶠᶜᶜ, δxᶠᶜᶠ, δxᶠᶠᶜ, δxᶠᶠᶠ,
+    δxᶜᶜᶜ, δxᶜᶜᶠ, δxᶜᶠᶜ, δxᶜᶠᶠ,
+    δyᶜᶠᶜ, δyᶜᶠᶠ, δyᶠᶠᶜ, δyᶠᶠᶠ,
+    δyᶜᶜᶜ, δyᶜᶜᶠ, δyᶠᶜᶜ, δyᶠᶜᶠ,
+    δzᶜᶜᶠ, δzᶜᶠᶠ, δzᶠᶜᶠ, δzᶠᶠᶠ,
+    δzᶜᶜᶜ, δzᶜᶠᶜ, δzᶠᶜᶜ, δzᶠᶠᶜ
 
 # Defining all the First order derivatives for the immersed boundaries
```
Comment (navidcy) [inline src/ImmersedBoundaries/conditional_derivatives.jl:9]: adjust the comment?

File: src/ImmersedBoundaries/conditional_derivatives.jl
Comment (navidcy) [inline src/ImmersedBoundaries/conditional_derivatives.jl:1]: rename the file to `conditional_differences.jl`?

**Review by navidcy — APPROVED** [review]

**Comment by navidcy** [comment]
The docs took 6hrs to build! Is the PR slowing down things? It could be just that tartarus was busy…

**Comment by simone-silvestri** [comment]
Oh wow! Let me check the performance

**Comment by glwagner** [comment]
A few tests are a good idea.

**Comment by simone-silvestri** [comment]
Indeed, the simulations seem to have slowed down significantly with these changes. I ll try to figure out why

**Comment by navidcy** [comment]
The alarms went off! (Do not merge label was added.)

:)

File: test/test_operators.jl [hunk 2]
```diff
@@ -5,6 +5,51 @@ using Oceananigans.Operators: Δyᵃᶠᵃ, Δyᵃᶜᵃ, Δyᶠᶠᵃ, Δyᶠ
...
+    for δy in (δyᶜᶠᶜ, δyᶠᶠᶜ, δyᶜᶠᶠ, δyᶠᶠᶠ)
+        assess = assess && δy(2, 2, 2, grid, f, ϕ) == δy_ϕ_f 
+    end
+
+    for δz in (δzᶜᶜᶜ, δzᶜᶠᶜ, δzᶠᶜᶜ, δzᶠᶠᶜ)
+        assess = assess && δz(2, 2, 2, grid, f, ϕ) == δz_ϕ_c 
+    end
+    for δz in (δzᶜᶜᶠ, δzᶜᶠᶠ, δzᶠᶜᶠ, δzᶠᶠᶠ)
+        assess = assess && δz(2, 2, 2, grid, f, ϕ) == δz_ϕ_f 
+    end
```
Comment (glwagner) [inline test/test_operators.jl:48]: Maybe make each of these a `@test`? It's nice to differentiate for debugging purposes

Comment (glwagner) [reply test/test_operators.jl:48]: If you need a variable like `assess`, a name like `operators_work` (which is `true` if they work, `false` if they don't) is a good readable option

File: src/Operators/difference_operators.jl [hunk 3]
```diff
@@ -47,3 +47,27 @@ using Oceananigans.Grids: Flat
...
+    δx = Symbol(:δx, ℓx, ℓy, ℓz)
+    δy = Symbol(:δy, ℓx, ℓy, ℓz)
+    δz = Symbol(:δz, ℓx, ℓy, ℓz)
+
+    δxᵃ = Symbol(:δx, ℓx, :ᵃ, :ᵃ)
+    δyᵃ = Symbol(:δy, :ᵃ, ℓy, :ᵃ)
+    δzᵃ = Symbol(:δz, :ᵃ, :ᵃ, ℓz)
+
+    @eval begin
+        $δx(i, j, k, grid, f::Function, args...) = $δxᵃ(i, j, k, grid, f, args...)
```
Comment (glwagner) [inline src/Operators/difference_operators.jl:65]:
```suggestion
        @inline $δx(i, j, k, grid, f::Function, args...) = $δxᵃ(i, j, k, grid, f, args...)
```

**Comment by simone-silvestri** [comment]
Should I merge? It looks like it is working

**Review by glwagner — APPROVED** [review]
Do we use functions to encapsulate a list of tests so that we can comment them out more easily? I don't think we do that elsewhere but it could be a good idea. Not sure.

**Comment by simone-silvestri** [comment]
I think it's also somewhere else, but cannot remember where

**Pull request merged**